File: src/com/l2kt/gameserver/data/xml/PlayerData.java
package com.l2kt.gameserver.data.xml;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.l2kt.commons.data.xml.IXmlReader;

import com.l2kt.gameserver.model.actor.template.PlayerTemplate;
import com.l2kt.gameserver.model.base.ClassId;
import com.l2kt.gameserver.model.holder.skillnode.GeneralSkillNode;
import com.l2kt.gameserver.model.location.Location;
import com.l2kt.gameserver.templates.StatsSet;

import org.w3c.dom.Document;

/**
 * This class loads and stores {@link PlayerTemplate}s. It also feed their skill trees.
 */
public class PlayerData implements IXmlReader
{
	private final Map<Integer, PlayerTemplate> _templates = new HashMap<>();
	
	protected PlayerData()
	{
		load();
	}
	
	@Override
	public void load()
	{
		parseFile("./data/xml/classes");
		LOGGER.info("Loaded {} player classes templates.", _templates.size());
		
		// We add parent skills, if existing.
		for (PlayerTemplate template : _templates.values())
		{
			final ClassId parentClassId = template.getClassId().getParent();
			if (parentClassId != null)
				template.getSkills().addAll(_templates.get(parentClassId.getId()).getSkills());
		}
	}
	
	@Override
	public void parseDocument(Document doc, Path path)
	{
		forEach(doc, "list", listNode -> forEach(listNode, "class", classNode ->
		{
			final StatsSet set = new StatsSet();
			forEach(classNode, "set", setNode -> set.putAll(parseAttributes(setNode)));
			forEach(classNode, "skills", skillsNode ->
			{
				final List<GeneralSkillNode> skills = new ArrayList<>();
				forEach(skillsNode, "skill", skillNode -> skills.add(new GeneralSkillNode(parseAttributes(skillNode))));
				set.set("skills", skills);
			});
			forEach(classNode, "spawns", spawnsNode ->
			{
				final List<Location> locs = new ArrayList<>();
				forEach(spawnsNode, "spawn", spawnNode -> locs.add(new Location(parseAttributes(spawnNode))));
				set.set("spawnLocations", locs);
			});
			_templates.put(set.getInteger("id"), new PlayerTemplate(set));
		}));
	}
	
	public PlayerTemplate getTemplate(ClassId classId)
	{
		return _templates.get(classId.getId());
	}
	
	public PlayerTemplate getTemplate(int classId)
	{
		return _templates.get(classId);
	}
	
	public final String getClassNameById(int classId)
	{
		final PlayerTemplate template = _templates.get(classId);
		return (template != null) ? template.getClassName() : "Invalid class";
	}
	
	public static PlayerData getInstance()
	{
		return SingletonHolder.INSTANCE;
	}
	
	private static class SingletonHolder
	{
		protected static final PlayerData INSTANCE = new PlayerData();
	}
}

File: src/com/l2kt/gameserver/data/xml/PlayerData.kt
package com.l2kt.gameserver.data.xml

import com.l2kt.commons.data.xml.IXmlReader
import com.l2kt.gameserver.data.xml.PlayerData.forEach
import com.l2kt.gameserver.model.actor.template.PlayerTemplate
import com.l2kt.gameserver.model.base.ClassId
import com.l2kt.gameserver.model.holder.skillnode.GeneralSkillNode
import com.l2kt.gameserver.model.location.Location
import com.l2kt.gameserver.templates.StatsSet
import org.w3c.dom.Document
import java.nio.file.Path
import java.util.*

/**
 * This class loads and stores [PlayerTemplate]s. It also feed their skill trees.
 */
object PlayerData : IXmlReader {
    private val _templates = HashMap<Int, PlayerTemplate>()

    init {
        load()
    }

    override fun load() {
        parseFile("./data/xml/classes")
        IXmlReader.LOGGER.info("Loaded {} player classes templates.", _templates.size)

        // We add parent skills, if existing.
        for (template in _templates.values) {
            val parentClassId = template.classId.parent
            if (parentClassId != null)
                template.skills.addAll(_templates[parentClassId.id]!!.skills)
        }
    }

    override fun parseDocument(doc: Document, path: Path) {
        forEach(doc, "list") { listNode ->
            forEach(listNode, "class") { classNode ->
                val set = StatsSet()
                forEach(classNode, "set") { setNode -> set.putAll(parseAttributes(setNode)) }
                forEach(classNode, "skills") { skillsNode ->
                    val skills = ArrayList<GeneralSkillNode>()
                    forEach(
                        skillsNode,
                        "skill"
                    ) { skillNode -> skills.add(GeneralSkillNode(parseAttributes(skillNode))) }
                    set.set("skills", skills)
                }
                forEach(classNode, "spawns") { spawnsNode ->
                    val locs = ArrayList<Location>()
                    forEach(spawnsNode, "spawn") { spawnNode -> locs.add(Location(parseAttributes(spawnNode))) }
                    set.set("spawnLocations", locs)
                }
                _templates[set.getInteger("id")] = PlayerTemplate(set)
            }
        }
    }

    fun getTemplate(classId: ClassId): PlayerTemplate? {
        return _templates[classId.id]
    }

    fun getTemplate(classId: Int): PlayerTemplate? {
        return _templates[classId]
    }

    fun getClassNameById(classId: Int): String {
        val template = _templates[classId]
        return if (template != null) template.className else "Invalid class"
    }
}